### PR TITLE
[JENKINS-63727] GitToolChooser should not recommend JGit if Git Publisher is present

### PR DESCRIPTION
**Pull request by rishabhBudhouliya**
## [JENKINS-63727](https://issues.jenkins-ci.org/browse/JENKINS-63727) - GitToolChooser should not suggest JGit as a viable git implementation if Git Publisher is chosen as a post build action

The current implementation of the GitToolChooser is not aware of the fact that JGit is not compatible with GitPublisher. To do so:
- Added a new UnsupportedCommand feature which would convey if JGit should be used in
the scenario where Git Publisher is being used
- Instantiated that command within the GitPublisher to pass down that info of compatibility
- Changed the CreateClient API to add UnsupportedCommand as a new parameter

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged in upstream modules (like git-client-plugin)
- [ ] Any dependent changes have been published in upstream modules (like git-client-plugin)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Further comments

This PR is dependent upon: https://github.com/jenkinsci/git-client-plugin/pull/615

### CODE REVIEW

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Thanks for working on this issue so quickly!  Much appreciated!

There is a compatibility break included in the proposed change that will need to be resolved before this can be merged.

File: src/test/java/hudson/plugins/git/GitPublisherTest.java
Comment (rishabhBudhouliya) [reply src/test/java/hudson/plugins/git/GitPublisherTest.java:143]: The only reason for choosing a matrix project was the previous test case. I chose that test case as a reference point to test the working of Git Publisher, it was the fastest way to do it for me. 

I should be able to replace it with a Freestyle project which understandably is a better choice.

File: src/main/java/hudson/plugins/git/GitSCM.java
Comment (rishabhBudhouliya) [reply src/main/java/hudson/plugins/git/GitSCM.java:827]: I don't think the existing signature should be deprecated since the argument, UnsupportedCommand, is only passed down by the Git Publisher. Hence it makes sense to retain the old signature method and add a new method which is used by the Git Publisher alone (earlier the CreateClient API was used commonly by both the SCM builder and Git Publisher)

File: pom.xml
Comment (MarkEWaite) [inline pom.xml:104]: Use the incremental build from git client plugin CI job.
```suggestion
      <version>3.5.1-rc2571.b85fc89d5173</version>
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
When I run the tests with this incremental build of the git client plugin master branch (including the new API),

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite I have added the requested changes, I am still not sure if we should deprecate the older CreateClient APIs considering the fact that the new signature method is only being used by the Git Publisher.


> When I run the tests with this incremental build of the git client plugin master branch (including the new API),

This comment seems incomplete for some reason.

**Comment by MarkEWaite** [comment]
> @MarkEWaite I have added the requested changes, I am still not sure if we should deprecate the older CreateClient APIs considering the fact that the new signature method is only being used by the Git Publisher.

Once we allow a public API into a release, we really don't know who is using it.  We rarely delete public API's because a deletion will break compatibility for consumers that we cannot identify.

The JGit 4.5.4 to JGit 5.x transition was a painful process because they deleted specific API's and we were intentionally bundling JGit.  Thankfully, they have not deleted API's since that transition.

> 
> > When I run the tests with this incremental build of the git client plugin master branch (including the new API),
> 
> This comment seems incomplete for some reason.

I should have finished the comment.  Running with the latest git client plugin incremental build, I was seeing local test failures that I could not explain.  I've not had time to investigate if those failures are consistent and repeatable or if there were purely transient.  Unfortunately, I probably won't have time to investigate the local failure.  My local failure does not appear on ci.jenkins.io, so I assume it is specific to my local environment.

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite Got it. If the jobs where you are seeing these failures are on your accessible local machine, I can try to check if those failures are transient or something we should worry about though the git plugin tests were running successfully on my local machine + ci.jenkins.io.

File: src/main/java/hudson/plugins/git/GitSCM.java
Comment (MarkEWaite) [inline src/main/java/hudson/plugins/git/GitSCM.java:890]: I needed to study this carefully before I understood what it is expressing.

As far as I can tell, it says that if the post build action does not support JGit or any of the extensions do not support JGit, then the chooser will not choose JGit.

Could the variable `cmd` be named more clearly as `postBuildUnsupportedCommand` or something like that?  I was initially confused by the reference to `cmd.determineSupportForJGit` and to `unsupportedCommand.determineSupportForJGit`./

Optionally, I think it could be expressed as a single statement:

```suggestion
                chooser = new GitToolChooser(url, project, ucCredentialsId, gitTool, n, listener,
                                                                   cmd.determineSupportForJGit() && unsupportedCommand.determineSupportForJGit());
```

File: src/test/java/hudson/plugins/git/GitPublisherTest.java
Comment (MarkEWaite) [reply src/test/java/hudson/plugins/git/GitPublisherTest.java:143]: I'm fine with it being a matrix project.  Nice to have a different project type included in the tests.  No need to change it.

File: src/main/java/hudson/plugins/git/GitSCM.java
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/git/GitSCM.java:849]: Unless there is a compelling reason to make the new method public, I'd rather leave it package protected so that only callers from within the git plugin can use it.  If you decide to keep it public, then the javadoc for the additional parameter needs to be made more general purpose so that other callers understand why they must provide a value there and how they can decide the correct value.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
I'd like a clearer name for the `cmd` variable so that I understand in context that the `cmd` is provided from a post build action while the extensions are provided by scanning the GitSCMExtensions

**Review by MarkEWaite — APPROVED** [review]
My experiments and testing have found no reason to block this pull request.  I don't yet understand why the automated tests do not fail when the fix is removed  from the code, nor do I understand why the automated test that I added is able to use JGit to push a branch, yet the multibranch test reports it is pushing a tag, the build log says the tag is being pushed, there is no exception, and yet the destination repository does not include the tag.

More study is needed to understand why JGit is not working with GitPublisher, but that study should not block the release of this fix to users.

**Pull request merged**
